atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTableIntegrationTest.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.pue;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.palantir.atlasdb.transaction.encoding.TwoPhaseEncodingStrategy;
+import com.palantir.common.concurrent.PTExecutors;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.junit.Test;
+
+public class ResilientCommitTimestampPutUnlessExistsTableIntegrationTest {
+    private static final double WRITE_FAILURE_PROBABILITY = 0.5;
+    private static final long MAXIMUM_EVALUATED_TIMESTAMP = 100;
+
+    private final ConsensusForgettingStore forgettingStore =
+            new CassandraImitatingConsensusForgettingStore(WRITE_FAILURE_PROBABILITY);
+    private final PutUnlessExistsTable<Long, Long> pueTable =
+            new ResilientCommitTimestampPutUnlessExistsTable(forgettingStore, TwoPhaseEncodingStrategy.INSTANCE);
+
+    @Test
+    public void repeatableReads() throws InterruptedException {
+        Multimap<Long, TimestampReader> timestampReaders = createStartedTimestampReaders();
+
+        CountDownLatch writeExecutionLatch = new CountDownLatch(1);
+        ExecutorService writeExecutor = Executors.newCachedThreadPool();
+
+        for (long startTimestamp = 1; startTimestamp <= MAXIMUM_EVALUATED_TIMESTAMP; startTimestamp++) {
+            for (int concurrentWriter = 1; concurrentWriter <= 10; concurrentWriter++) {
+                int finalConcurrentWriter = concurrentWriter;
+                long finalStartTimestamp = startTimestamp;
```

Comment:
You can extract a method to avoid having to do this?

Reply:
Yes. Done